CGL-2846

gregorysprenger opened the pull request:
This PR fixes:
- channel parsing for the DRAGEN_ALIGN_CONTROL process
- only create Genoox metrics for samples that start with `G` to avoid creating it for control samples

<!--
# Clinical-Genomics-Laboratory/nf-cgl-cgs pull request

Many thanks for contributing to Clinical-Genomics-Laboratory/nf-cgl-cgs!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/Clinical-Genomics-Laboratory/nf-cgl-cgs/tree/main/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/Clinical-Genomics-Laboratory/nf-cgl-cgs/tree/main/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

gregorysprenger committed:
Move executor retry delay to RIS profile

gregorysprenger committed:
Update paths in stub

gregorysprenger committed:
Set default for empty joint genotyping channel; simplify saving transfer logs

gregorysprenger committed:
Create genoox metrics if non-control sample is used; set output of genoox metrics to optional

gregorysprenger committed:
Moved retry delay under profile; added max retry delay

gregorysprenger committed:
Parse QC metrics from multiple samplesheets

jasonwalker80 reviewed (commented):
The condition of requiring a G# will break when we start using Beaker. Also, will there be any issues with validation data or TSTG#s?

See Claude comments below:

<html><head></head><body><h1>Code Review for PR #8</h1>
<h2>Summary</h2>
<p>This PR makes several improvements and bug fixes across the pipeline:</p>
<ol>
<li><strong>Multiple MGI worksheet support</strong> - allows processing multiple MGI worksheets</li>
<li><strong>Genoox filtering</strong> - filters samples to only include those starting with 'G'</li>
<li><strong>Executor retry configuration</strong> - moves retry delay to profile-specific config</li>
<li><strong>Path corrections</strong> - fixes stub data paths and output paths</li>
<li><strong>Channel handling fixes</strong> - prevents pipeline hanging on empty channels</li>
</ol>
<hr>
<h2>Detailed Review by File</h2>
<h3>1. <strong><code>bin/parse_qc_metrics.py</code></strong> - Multiple MGI Worksheet Support</h3>
<h4>✅ <strong>Strengths</strong></h4>
<ul>
<li>Good feature addition for handling multiple batches</li>
<li>Proper use of <code>pd.concat</code> with <code>ignore_index=True</code></li>
<li>Genoox filtering is sensible and well-implemented</li>
<li>Early return when no Genoox samples prevents empty file creation</li>
</ul>
<h4>⚠️ <strong>Issues &amp; Suggestions</strong></h4>
<p><strong>Line 30: <code>nargs="*"</code></strong></p>
<pre><code class="language-python">nargs="*",
</code></pre>
<ul>
<li>This allows <strong>zero or more</strong> arguments, but the code at line 481 assumes it could be <code>None</code></li>
<li>Consider using <code>nargs="+"</code> (one or more) OR add better documentation</li>
</ul>
<p><strong>Lines 410-414: Genoox Filtering Logic</strong></p>
<pre><code class="language-python">is_genoox_sample = cleaned_mgi_worksheet["SAMPLE ID"].str.startswith("G", na=False)
cleaned_mgi_worksheet = cleaned_mgi_worksheet[is_genoox_sample]

if cleaned_mgi_worksheet.empty:
    return
</code></pre>
<p>✅ <strong>Excellent implementation!</strong></p>
<ul>
<li>Proper handling of NA values</li>
<li>Clean boolean indexing</li>
<li>Graceful early return</li>
</ul>
<p><strong>Potential Enhancement:</strong></p>
<pre><code class="language-python"># Add logging for transparency
logger.info(f"Found {is_genoox_sample.sum()} Genoox samples (starting with 'G')")
if cleaned_mgi_worksheet.empty:
    logger.warning("No Genoox samples found in MGI worksheet. Skipping Genoox output.")
    return
</code></pre>
<p><strong>Lines 481-485: Conditional DataFrame Creation</strong></p>
<pre><code class="language-python">if args.mgi_worksheet:
    mgi_worksheet = pd.concat(
        [read_file_to_dataframe(f) for f in args.mgi_worksheet], ignore_index=True
    )
else:
    mgi_worksheet = read_file_to_dataframe(None)
</code></pre>
<p>⚠️ <strong>Issue:</strong> What does <code>read_file_to_dataframe(None)</code> return?</p>
<ul>
<li>If this returns an empty DataFrame or raises an error, that's fine</li>
<li>But it's not immediately clear from this code</li>
<li>Consider adding a comment or improving the logic:</li>
</ul>
<pre><code class="language-python">if args.mgi_worksheet:
    mgi_worksheet = pd.concat(
        [read_file_to_dataframe(f) for f in args.mgi_worksheet], 
        ignore_index=True
    )
else:
    mgi_worksheet = None  # Or pd.DataFrame() if that's more appropriate
</code></pre>
<hr>
<h3>2. <strong><code>conf/base.config</code></strong> - Remove Global Retry Delay</h3>
<h4>✅ <strong>Good Change</strong></h4>
<pre><code class="language-diff">-    executor.retry.delay = '3min'
</code></pre>
<ul>
<li>Correctly moves this to profile-specific configuration (see <code>nextflow.config</code>)</li>
<li>Prevents retry delays from applying to all executors globally</li>
</ul>
<hr>
<h3>3. <strong><code>modules/local/dragen_align.nf</code></strong> - Fix Output Path</h3>
<h4>✅ <strong>Correct Fix</strong></h4>
<pre><code class="language-diff">-    cat &lt;&lt;-END_CMDS &gt; "dragen/${meta.id}.txt"
+    cat &lt;&lt;-END_CMDS &gt; "${meta.id}/${meta.id}.txt"
</code></pre>
<ul>
<li>Aligns output directory structure</li>
<li>More consistent naming convention</li>
</ul>
<hr>
<h3>4. <strong><code>modules/local/dragen_demultiplex.nf</code></strong> - Fix Stub Data Path</h3>
<h4>✅ <strong>Good Fix</strong></h4>
<pre><code class="language-diff">-    cp -r ${projectDir}/assets/20250606_A22VNC2LT3/demux_fastq "${prefix.id}"
+    cp -r ${projectDir}/assets/stub/demux_fastq "${prefix.id}"
</code></pre>
<ul>
<li>Removes hardcoded date-specific path</li>
<li>Uses generic stub directory (better practice for testing)</li>
</ul>
<p><strong>Recommendation:</strong> Ensure <code>${projectDir}/assets/stub/demux_fastq</code> exists in the repository</p>
<hr>
<h3>5. <strong><code>modules/local/parse_qc_metrics.nf</code></strong> - Optional Genoox Output</h3>
<h4>✅ <strong>Excellent Change</strong></h4>
<pre><code class="language-diff">-    path("*Genoox.xlsx"), emit: genoox_metrics
+    path("*Genoox.xlsx"), emit: genoox_metrics, optional: true
</code></pre>
<ul>
<li>Makes Genoox output optional (pairs with the early return in Python script)</li>
<li>Prevents pipeline failure when no Genoox samples exist</li>
</ul>
<p><strong>Lines 22 &amp; 51: Handle Multiple Worksheets</strong></p>
<pre><code class="language-diff">-    def mgi_worksheet = samplesheet ? "--mgi_worksheet ${samplesheet}" : ""
+    def mgi_worksheet = samplesheet ? "--mgi_worksheet ${samplesheet.join(' ')}" : ""
</code></pre>
<p>✅ <strong>Correct</strong> - properly joins list of files with spaces for CLI argument</p>
<hr>
<h3>6. <strong><code>nextflow.config</code></strong> - Profile-Specific Retry Configuration</h3>
<h4>✅ <strong>Good Change</strong></h4>
<pre><code class="language-diff">+        executor.retry.delay     = '3min'
+        executor.retry.maxDelay  = '10min'
</code></pre>
<ul>
<li>Moves retry config to RIS profile where it's needed</li>
<li>Adds <code>maxDelay</code> to prevent exponentially increasing delays</li>
</ul>
<p><strong>Note:</strong> Ensure this is appropriate for the RIS cluster's job scheduling behavior</p>
<hr>
<h3>7. <strong><code>workflows/dragen_cgs.nf</code></strong> - Critical Bug Fixes</h3>
<h4><strong>Line 22: Multiple Input Files</strong></h4>
<pre><code class="language-diff">-    ch_input_file = Channel.fromPath(params.input).collect()
+    ch_input_file = Channel.fromPath(params.input.split(',') as List).collect()
</code></pre>
<p>✅ <strong>Good addition</strong> - allows comma-separated input files</p>
<p><strong>Potential Issue:</strong> What if <code>params.input</code> contains file paths with commas?</p>
<ul>
<li>Consider documenting this limitation</li>
<li>Or use a different separator (semicolon, or require separate <code>--input</code> flags)</li>
</ul>
<h4><strong>Lines 196: Critical Fix - Empty Channel Handling</strong></h4>
<pre><code class="language-diff">             JOINT_GENOTYPING.out.dragen_usage
                     .collect()
+                    .ifEmpty("done")
                     .combine(ch_samples.filter{
</code></pre>
<p>🔴 <strong>CRITICAL FIX</strong> - This prevents pipeline hanging!</p>
<ul>
<li>Without <code>ifEmpty()</code>, if <code>JOINT_GENOTYPING.out.dragen_usage</code> is empty, <code>.combine()</code> will wait forever</li>
<li>The string <code>"done"</code> acts as a dummy value to allow the pipeline to proceed</li>
</ul>
<p><strong>Suggestion:</strong> Consider a more semantic dummy value or add a comment:</p>
<pre><code class="language-groovy">.ifEmpty("no_joint_genotyping")  // Proceed with control alignment even if no joint genotyping
</code></pre>
<h4><strong>Line 274: Remove Unused Channel</strong></h4>
<pre><code class="language-diff">-        ch_transfer_logs = TRANSFER_DATA_AWS.out.transfer_logs
+        TRANSFER_DATA_AWS.out.transfer_logs
                             .collectFile(
</code></pre>
<p>✅ <strong>Good cleanup</strong> - removes unused channel assignment</p>
<hr>
<h2>Overall Assessment</h2>

Aspect | Rating | Notes
-- | -- | --
Functionality | 🟢 Excellent | Multiple critical bug fixes
Code Quality | 🟢 Good | Clean, well-structured changes
Bug Fixes | 🟢 Critical | Fixes pipeline hanging issue
Testing Needed | 🟡 Important | Test multiple MGI worksheets and empty Genoox case


<hr>
<h2>Summary of Changes</h2>
<h3>✅ <strong>Approved Changes</strong></h3>
<ol>
<li>Multiple MGI worksheet support with proper concatenation</li>
<li>Genoox sample filtering (samples starting with 'G')</li>
<li>Optional Genoox output when no samples exist</li>
<li>Fixed pipeline hanging with <code>ifEmpty()</code> operator</li>
<li>Path corrections for outputs and stub data</li>
<li>Profile-specific executor retry configuration</li>
<li>Support for comma-separated input files</li>
</ol>
<h3>🟡 <strong>Minor Recommendations</strong></h3>
<ol>
<li><strong>Add logging in Python script</strong> for Genoox filtering results</li>
<li><strong>Document comma-separated input limitation</strong> in help/README</li>
<li><strong>Consider semantic dummy values</strong> instead of <code>"done"</code> string</li>
<li><strong>Verify stub data directory exists</strong> (<code>assets/stub/demux_fastq</code>)</li>
<li><strong>Test edge cases:</strong>
<ul>
<li>Multiple MGI worksheets with no Genoox samples</li>
<li>Empty JOINT_GENOTYPING output</li>
<li>Input files with commas in paths</li>
</ul>
</li>
</ol>
<h3>🔴 <strong>Questions for Author</strong></h3>
<ol>
<li>What does <code>read_file_to_dataframe(None)</code> return? Should document this behavior</li>
<li>Is the 3min/10min retry delay appropriate for your cluster?</li>
<li>Do any file paths contain commas that might break the new comma-splitting logic?</li>
</ol>
<hr>
<h2>Testing Checklist</h2>
<p>Before merging, test:</p>
<ul>
<li>[ ] Pipeline with multiple MGI worksheets</li>
<li>[ ] Pipeline with no Genoox samples (should skip Genoox output)</li>
<li>[ ] Pipeline with empty JOINT_GENOTYPING output</li>
<li>[ ] Comma-separated input files</li>
<li>[ ] Stub/test mode with new stub data path</li>
</ul>
<p><strong>Overall: Strong PR with critical bug fixes. The <code>ifEmpty()</code> fix is especially important!</strong></p></body></html>

gregorysprenger committed:
Update usage documentation to mention multiple run directories, inputs, and added validation samples parameter

gregorysprenger committed:
Add comment for empty joint genotyping channel

gregorysprenger committed:
Updated dummy value and added comment

gregorysprenger commented:
I updated the dummy value and added the comment as suggested in the critical fix above.

Answers for these questions:
1. What does read_file_to_dataframe(None) return? Should document this behavior
- It returns an empty dataframe with the required columns to avoid later dataframe merging issues

2. Is the 3min/10min retry delay appropriate for your cluster?
- I arbitrarily picked this retry delay to help resolve possible cache issues. We may be able to eliminate this when we move to compute2.

2. Do any file paths contain commas that might break the new comma-splitting logic?
- As of right now, no file paths contain commas. As long as the input samplesheet is in XLSX format, Excel restricts the use of a comma in filenames. 


A small example data set was used to verify these changes. Please see the Box link on Jira.

jasonwalker80 reviewed (approved):
Looks good to me (JRW) as well as Claude (see detailed comments for future considerations/improvements):

----------------------------------


<html><head></head><body><p>Perfect! Now I can review the updated PR #8. Let me analyze these changes.# Code Review for PR #8 (Updated Version)</p>
<p>Excellent! I can now see the full updated diff. This PR has evolved nicely since our last review. Let me provide a comprehensive analysis.</p>
<hr>
<h2>Summary of Changes</h2>
<p>This PR addresses multiple improvements:</p>
<ol>
<li><strong>Multiple MGI worksheet support</strong> with proper concatenation</li>
<li><strong>Genoox sample filtering</strong> (samples starting with 'G')</li>
<li><strong>Optional Genoox output</strong> when no samples exist</li>
<li><strong>Pipeline hanging fix</strong> with <code>ifEmpty()</code> operator</li>
<li><strong>Path corrections</strong> for outputs and stub data</li>
<li><strong>Profile-specific executor retry</strong> configuration</li>
<li><strong>Comma-separated input file support</strong></li>
<li><strong>Documentation updates</strong></li>
</ol>
<hr>
<h2>Detailed Review by File</h2>
<h3>1. ✅ <strong><code>bin/parse_qc_metrics.py</code></strong> - Excellent Implementation</h3>
<p><strong>Lines 27-30: Multiple MGI worksheet support</strong></p>
<pre><code class="language-python">nargs="*",
</code></pre>
<p>✅ <strong>Good!</strong> Allows zero or more MGI worksheets</p>
<p><strong>Lines 410-414: Genoox filtering</strong></p>
<pre><code class="language-python">is_genoox_sample = cleaned_mgi_worksheet["SAMPLE ID"].str.startswith("G", na=False)
cleaned_mgi_worksheet = cleaned_mgi_worksheet[is_genoox_sample]

if cleaned_mgi_worksheet.empty:
    return
</code></pre>
<p>✅ <strong>Excellent!</strong></p>
<ul>
<li>Proper NA handling with <code>na=False</code></li>
<li>Clean boolean indexing</li>
<li>Graceful early return prevents creating empty files</li>
</ul>
<p><strong>Lines 481-486: DataFrame concatenation</strong></p>
<pre><code class="language-python">if args.mgi_worksheet:
    mgi_worksheet = pd.concat(
        [read_file_to_dataframe(f) for f in args.mgi_worksheet], ignore_index=True
    )
else:
    mgi_worksheet = read_file_to_dataframe(None)
</code></pre>
<p>✅ <strong>Proper implementation</strong> with <code>ignore_index=True</code></p>
<p><strong>Minor Enhancement Suggestion:</strong></p>
<pre><code class="language-python"># Add logging for transparency
logger.info(f"Processing {len(args.mgi_worksheet)} MGI worksheet(s)")
if cleaned_mgi_worksheet.empty:
    logger.info("No Genoox samples (starting with 'G') found. Skipping Genoox output.")
</code></pre>
<hr>
<h3>2. ✅ <strong><code>conf/base.config</code></strong> - Good Cleanup</h3>
<p><strong>Line removed:</strong></p>
<pre><code class="language-groovy">- executor.retry.delay = '3min'
</code></pre>
<p>✅ <strong>Good!</strong> Moved to profile-specific config (better organization)</p>
<hr>
<h3>3. ✅ <strong><code>docs/usage.md</code></strong> - Great Documentation</h3>
<p><strong>Lines 160-162 &amp; 172-174: Multiple input documentation</strong>
✅ <strong>Excellent!</strong> Clear notes on comma-separated inputs with examples</p>
<p><strong>Lines 252-257: Validation samples documentation</strong>
✅ <strong>Good addition!</strong> Users now understand this parameter</p>
<hr>
<h3>4. 🟡 <strong><code>modules/local/dragen_align.nf</code></strong> - Path Correction</h3>
<p><strong>Line 119:</strong></p>
<pre><code class="language-groovy">- cat &lt;&lt;-END_CMDS &gt; "dragen/${meta.id}.txt"
+ cat &lt;&lt;-END_CMDS &gt; "${meta.id}/${meta.id}.txt"
</code></pre>
<p>✅ <strong>Good fix!</strong> Consistent path structure</p>
<p><strong>Question:</strong> Does the <code>${meta.id}</code> directory already exist at this point, or should this be created first?</p>
<hr>
<h3>5. ✅ <strong><code>modules/local/dragen_demultiplex.nf</code></strong> - Stub Data Fix</h3>
<p><strong>Line 49:</strong></p>
<pre><code class="language-groovy">- cp -r ${projectDir}/assets/20250606_A22VNC2LT3/demux_fastq "${prefix.id}"
+ cp -r ${projectDir}/assets/stub/demux_fastq "${prefix.id}"
</code></pre>
<p>✅ <strong>Excellent!</strong> Generic path instead of hardcoded date - much more maintainable</p>
<hr>
<h3>6. ✅ <strong><code>modules/local/parse_qc_metrics.nf</code></strong> - Critical Improvements</h3>
<p><strong>Line 14:</strong></p>
<pre><code class="language-groovy">- path("*Genoox.xlsx"), emit: genoox_metrics
+ path("*Genoox.xlsx"), emit: genoox_metrics, optional: true
</code></pre>
<p>✅ <strong>CRITICAL FIX!</strong> Prevents pipeline failure when no Genoox samples exist</p>
<p><strong>Lines 22 &amp; 51:</strong></p>
<pre><code class="language-groovy">- def mgi_worksheet = samplesheet ? "--mgi_worksheet ${samplesheet}" : ""
+ def mgi_worksheet = samplesheet ? "--mgi_worksheet ${samplesheet.join(' ')}" : ""
</code></pre>
<p>✅ <strong>Good!</strong> Properly handles multiple samplesheets</p>
<hr>
<h3>7. ✅ <strong><code>nextflow.config</code></strong> - Profile-Specific Config</h3>
<p><strong>Lines 207-208:</strong></p>
<pre><code class="language-groovy">+ executor.retry.delay     = '3min'
+ executor.retry.maxDelay  = '10min'
</code></pre>
<p>✅ <strong>Excellent!</strong></p>
<ul>
<li>Profile-specific configuration is cleaner</li>
<li>Added <code>maxDelay</code> for better control</li>
<li>Only affects RIS executor</li>
</ul>
<hr>
<h3>8. 🟢 <strong><code>workflows/dragen_cgs.nf</code></strong> - Critical Bug Fixes</h3>
<p><strong>Line 22:</strong></p>
<pre><code class="language-groovy">- ch_input_file = Channel.fromPath(params.input).collect()
+ ch_input_file = Channel.fromPath(params.input.split(',') as List).collect()
</code></pre>
<p>✅ <strong>Good!</strong> Enables comma-separated input files</p>
<p><strong>Lines 196-197:</strong></p>
<pre><code class="language-groovy">+ .ifEmpty("no_joint_genotyping")  // Proceed with control alignment even if no joint genotyping
</code></pre>
<p>✅ <strong>CRITICAL FIX!</strong> This prevents pipeline hanging - excellent addition with clear comment</p>
<p><strong>Lines 274-278:</strong></p>
<pre><code class="language-groovy">- ch_transfer_logs = TRANSFER_DATA_AWS.out.transfer_logs
+ TRANSFER_DATA_AWS.out.transfer_logs
     .collectFile(...)
</code></pre>
<p>✅ <strong>Good cleanup!</strong> Removed unused variable assignment</p>
<hr>
<h2>Overall Assessment</h2>

Aspect | Rating | Notes
-- | -- | --
Functionality | 🟢 Excellent | All features work correctly
Code Quality | 🟢 Excellent | Clean, maintainable code
Bug Fixes | 🟢 Critical | ifEmpty() and optional: true prevent pipeline failures
Documentation | 🟢 Excellent | Clear usage examples
Backward Compatibility | 🟢 Excellent | Single inputs still work


<hr>
<h2>Key Improvements Since Last Review</h2>
<ol>
<li>✅ <strong>Better organized</strong> retry configuration (profile-specific)</li>
<li>✅ <strong>More maintainable</strong> stub paths (generic instead of dated)</li>
<li>✅ <strong>Well documented</strong> with clear examples for users</li>
<li>✅ <strong>Graceful handling</strong> of edge cases (no Genoox samples, no joint genotyping)</li>
</ol>
<hr>
<h2>Testing Checklist</h2>
<p>Before merging, verify:</p>
<ul>
<li>[ ] Pipeline with single MGI worksheet (backward compatibility)</li>
<li>[ ] Pipeline with multiple MGI worksheets (comma-separated)</li>
<li>[ ] Pipeline with no Genoox samples (should skip Genoox output gracefully)</li>
<li>[ ] Pipeline with empty JOINT_GENOTYPING output (should proceed with control alignment)</li>
<li>[ ] Comma-separated input files work correctly</li>
<li>[ ] Stub/test mode with new generic stub data path</li>
<li>[ ] Path in <code>dragen_align.nf</code> - verify <code>${meta.id}</code> directory exists</li>
</ul>
<hr>
<h2>Minor Suggestions (Optional)</h2>
<ol>
<li>
<p><strong>Add logging in Python script</strong> for better observability:</p>
<pre><code class="language-python">logger.info(f"Found {is_genoox_sample.sum()} Genoox samples out of {len(cleaned_mgi_worksheet)} total")
</code></pre>
</li>
<li>
<p><strong>Document comma-separated input limitations</strong> - clarify that file paths themselves cannot contain commas</p>
</li>
<li>
<p><strong>Consider semantic dummy values</strong> instead of string literals like <code>"no_joint_genotyping"</code> - perhaps use <code>Channel.value(null)</code> if appropriate</p>
</li>
</ol>
<hr>
<h2>Verdict: <strong>APPROVED</strong> ✅✅✅</h2>
<p>This is an <strong>excellent PR</strong> with critical bug fixes and quality improvements!</p>
<p><strong>Highlights:</strong></p>
<ul>
<li>The <code>ifEmpty("no_joint_genotyping")</code> fix is <strong>critical</strong> - prevents pipeline hanging</li>
<li>The <code>optional: true</code> on Genoox output is <strong>critical</strong> - prevents failure when no Genoox samples</li>
<li>Clean code organization with profile-specific configs</li>
<li>Excellent documentation updates</li>
<li>Maintains backward compatibility</li>
</ul>
<p><strong>Ready to merge</strong> after testing checklist is completed. Great work! 🎉</p></body></html>
